### DISCUSSION TITLE
🎉 Pool optimizer testing party / bug hunt 🐛

### DISCUSSION THREAD

**Discussion by Seldaek** [composer/composer#10276] [Show and tell]
## Update Nov 23rd: Optimizer is now enabled by default

Edit: The latest snapshot now has pool optimizer enabled by default, you can disable it with `COMPOSER_POOL_OPTIMIZER=0` if you think something is wrong, but hopefully this won't be needed.

Thanks everyone for helping out here!

----

https://github.com/composer/composer/pull/9261 introduces a pool optimizer (thanks @Toflar!). This adds an additional step when running updates (and `require`) between the metadata retrieval and the dependency resolution, filtering package versions before the dependency solver gets to look at them to reduce both memory usage and runtime.

Initial testing looks promising and it's hopefully conservative enough to be bug-free, but we could use help to test this assumption with a broader variety of projects and dependency setups.

## How can you help?

1. Make sure you run on the latest snapshot build: `composer self-update --snapshot`
2. To activate the optimizer, you need to set the `COMPOSER_POOL_OPTIMIZER` env var to `1`. You can do that by prepending the command with `COMPOSER_POOL_OPTIMIZER=1` or doing `export COMPOSER_POOL_OPTIMIZER=1` if you want it to persist throughout multiple commands as long as your terminal is open.
3. Use `-vv` to enable very verbose output, which will show you some debug output from this test.
4. Run an update command, either partial or full update, or a require command. If you're scared of the effects don't forget to use `--dry-run`.

TL;DR: `COMPOSER_POOL_OPTIMIZER=1 composer update --dry-run -vv` is a quick and safe way to try this.

You should then see some results without optimizer (above) and optimized pool (below), highlighted with those messages printed in red:

![image](https://user-images.githubusercontent.com/183678/141325830-b9d84239-bb63-42f6-8a75-c327046fa3c5.png)

If the results match it will print `Done, test successful` and continue, otherwise it should throw an exception.

Now please report your results as reactions on this post:

- If you see positive results but it was already fast enough please hit 👍🏻 
- If you see positive results and it improved speed a lot please hit 🚀 
- If you get an error please report it below as a [new issue](https://github.com/composer/composer/issues/new) with your composer.json so we can hopefully reproduce and fix it. You get to hit ❤️ because you found a bug.
- If you are too lazy to test or Composer is just great as it is and you don't care you can hit 🎉 anyway just to join in on the fun.

### Running only with the optimized pool

If you're happy and want to keep using it in fast mode only and not run with and without optimizer every time (we will enable it by default soon anyway if all goes well), you can set `COMPOSER_POOL_OPTIMIZER=2` to bypass the test mode. This does not help us validate that the behavior is still the same though.

### Thanks!

**Comment by bizon1** [comment]
Awesome guys, thank you for this great work.
Fast, fast, faster with no errors!
I tested at my host on uberspace
`[user@asteroid html]$ COMPOSER_POOL_OPTIMIZER=2 composer update -vv
Loading composer repositories with package information
Found 16,509 package versions referenced in your dependency graph. 13,725 (83%) were optimized away.
Updating dependencies
**Dependency resolution completed in 0.105 seconds**
Analyzed 2784 packages to resolve dependencies
Analyzed 78498 rules to resolve dependencies
Lock file operations: 2 installs, 104 updates, 10 removals`

**Comment by goetas** [comment]
i have this results 

```
Loading composer repositories with package information
> pre-pool-create: Symfony\Flex\Flex->truncatePackages
Restricting packages listed in "symfony/symfony" to "5.*"
Updating dependencies
Updating dependencies with default package pool
Dependency resolution completed in 1.574 seconds
Analyzed 5925 packages to resolve dependencies
Analyzed 44951 rules to resolve dependencies
Updating dependencies with optimized package pool
> pre-pool-create: Symfony\Flex\Flex->truncatePackages
Found 5,925 package versions referenced in your dependency graph. 4,458 (75%) were optimized away.
Dependency resolution completed in 0.168 seconds
Analyzed 1467 packages to resolve dependencies
Analyzed 9111 rules to resolve dependencies
Done, test successful
Dependency resolution completed in 0.005 seconds
Lock file operations: 0 installs, 54 updates, 0 removals


```

**Reply by dhrrgn** [reply]
Ran into an issue:

After updating to the snapshot, anytime I run any composer command, it is loading files listed in `autoload.files`, and is producing errors like `Class Foo\Bar not found`. 

This happens whether the dependency needed in that file exists or not. 

I have composer installed globally, not sure if that matters.

Let me know if you need more info. I can dig in further.

**Reply by Seldaek** [reply]
@dhrrgn please make sure you run latest snapshot and if you can still repro report a new issue with details.

**Comment by HenkPoley** [comment]
I'd like to add that in PowerShell the environment variable setting notation is `$Env:COMPOSER_POOL_OPTIMIZER="1";`

Works for me,  0.828 seconds -> 0.028 seconds:

```
> $Env:COMPOSER_POOL_OPTIMIZER="1"; C:\php-7.2.34-Win32-VC15-x64\php.exe .\composer.pool_optimizer.phar update --dry-run -vv
Loading composer repositories with package information
Updating dependencies
Updating dependencies with default package pool
Dependency resolution completed in 0.828 seconds
Analyzed 14697 packages to resolve dependencies
Analyzed 364450 rules to resolve dependencies
Updating dependencies with optimized package pool
Found 14,697 package versions referenced in your dependency graph. 12,886 (88%) were optimized away.
Dependency resolution completed in 0.028 seconds
Analyzed 1811 packages to resolve dependencies
Analyzed 15052 rules to resolve dependencies
Done, test successful
Nothing to modify in lock file
Dependency resolution completed in 0.000 seconds
Installing dependencies from lock file (including require-dev)
Nothing to install, update or remove
Package phpunit/php-token-stream is abandoned, you should avoid using it. No replacement was suggested.
Package webmozart/path-util is abandoned, you should avoid using it. Use symfony/filesystem instead.
107 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
```

Tests for our product also ran OK 👌

**Comment by Ocramius** [comment]
<details>
<summary>Ran a local test to see if `roave/security-advisories` would regress somehow - no visible changes on an artificial setup, and all good :heavy_check_mark:


No major performance issues, but no regressions either :) </summary>

```json
{
  "require": {
    "roave/security-advisories": "dev-latest",
    "doctrine/dbal": "^3,<3.1.3"
  }
}
```

```sh
❯ ./composer.phar --version
Composer version 2.2-dev (2.2-dev+8386264c9820e35d95d0df5f6a38727a6d1fe752) 2021-11-17 19:54:18

❯ time ./composer.phar update --dry-run -vv
Loading composer repositories with package information
Updating dependencies
Dependency resolution completed in 0.004 seconds
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - roave/security-advisories dev-latest conflicts with roave/security-advisories dev-latest.
    - Root composer.json requires roave/security-advisories dev-latest -> satisfiable by roave/security-advisories[dev-latest].
    - Root composer.json requires doctrine/dbal ^3,<3.1.3 -> satisfiable by doctrine/dbal[3.0.0, 3.1.0, 3.1.1, 3.1.2].


real	0m0,559s
user	0m0,290s
sys	0m0,112s

❯ time COMPOSER_POOL_OPTIMIZER=1 ./composer.phar update --dry-run -vv
Loading composer repositories with package information
Updating dependencies
Updating dependencies with default package pool
Dependency resolution completed in 0.002 seconds
Found 141 package versions referenced in your dependency graph. 37 (26%) were optimized away.
Dependency resolution completed in 0.000 seconds
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - roave/security-advisories dev-latest conflicts with roave/security-advisories dev-latest.
    - Root composer.json requires roave/security-advisories dev-latest -> satisfiable by roave/security-advisories[dev-latest].
    - Root composer.json requires doctrine/dbal ^3,<3.1.3 -> satisfiable by doctrine/dbal[3.0.0, 3.1.0, 3.1.1, 3.1.2].


real	0m0,481s
user	0m0,286s
sys	0m0,054s

❯ time COMPOSER_POOL_OPTIMIZER=2 ./composer.phar update --dry-run -vv
Loading composer repositories with package information
Found 141 package versions referenced in your dependency graph. 37 (26%) were optimized away.
Updating dependencies
Dependency resolution completed in 0.001 seconds
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - roave/security-advisories dev-latest conflicts with roave/security-advisories dev-latest.
    - Root composer.json requires roave/security-advisories dev-latest -> satisfiable by roave/security-advisories[dev-latest].
    - Root composer.json requires doctrine/dbal ^3,<3.1.3 -> satisfiable by doctrine/dbal[3.0.0, 3.1.0, 3.1.1, 3.1.2].


real	0m0,401s
user	0m0,219s
sys	0m0,075s
```

</details>

**Reply by stof** [reply]
if you want to see the performance improvement of future Composer versions when using `time` for benchmark, you need to use `COMPOSER_POOL_OPTIMIZER=2`.

With `COMPOSER_POOL_OPTIMIZER=1`, composer performs the dependency resolution twice to compare them.

**Reply by Ocramius** [reply]
Thanks, updated :)

**Comment by johandenhollander** [comment]
Just ran this a couple of times on a Drupal project:

Dependency resolution completed in 2.874 seconds
Total composer update time: 18s

Dependency resolution completed in 3.010 seconds
Total composer update time: 16s

********************************************************************************

Upgrading to version 8386264c9820e35d95d0df5f6a38727a6d1fe752 (snapshot channel).
Use composer self-update --rollback to return to version 2.0.11

********************************************************************************
COMPOSER_POOL_OPTIMIZER=1

Dependency resolution completed in 2.789 seconds
Total composer update time: 28s

Dependency resolution completed in 2.925 seconds
Total composer update time: 17 s

**Comment by Kingdutch** [comment]
I don't want to pollute this with success stories for the awesome reaction system you've set-up. Though I'm sorry, I can't not share my amazement at this awesome speed improvement! 🤯 

Time to add `export COMPOSER_POOL_OPTIMIZER=2` to my zsh startup. This is going to save so many developers  so much time ❤️ 

```
Loading composer repositories with package information
https://repo.packagist.com/<org> could not be fully loaded (curl error 55 while downloading https://repo.packagist.com/<org>/p2/npm-asset/graceful-fs~dev.json: Connection died, tried 5 times before giving up), package information was loaded from the local cache and may be out of date
Updating dependencies
Updating dependencies with default package pool
Dependency resolution completed in 218.508 seconds
Analyzed 24723 packages to resolve dependencies
Analyzed 2304743 rules to resolve dependencies
Updating dependencies with optimized package pool
Found 24,723 package versions referenced in your dependency graph. 20,195 (82%) were optimized away.
Dependency resolution completed in 0.724 seconds
Analyzed 4528 packages to resolve dependencies
Analyzed 41143 rules to resolve dependencies
Done, test successful
Dependency resolution completed in 0.002 seconds
Lock file operations: 0 installs, 20 updates, 0 removals
```

**Reply by Seldaek** [reply]
The latest snapshot now has pool optimizer enabled by default, so no need to keep the env var around :)

**Comment by navneet0693** [comment]
Same as what @Kingdutch said sorry 🙈  but I am going to join this success stories thread.

`**Composer version 2.1.12 2021-11-09 16:02:04**`
`composer update -vv`

```
Dependency resolution completed in 2.606 seconds
Analyzed 23504 packages to resolve dependencies
Analyzed 2158176 rules to resolve dependencies
Nothing to modify in lock file
Dependency resolution completed in 0.001 seconds
```

`Upgrading to version 8386264c9820e35d95d0df5f6a38727a6d1fe752 (snapshot channel).`

`COMPOSER_POOL_OPTIMIZER=1 composer update -vv`

```
Updating dependencies
Updating dependencies with default package pool
Dependency resolution completed in 2.547 seconds
Analyzed 23504 packages to resolve dependencies
Analyzed 2158176 rules to resolve dependencies
Updating dependencies with optimized package pool
Found 23,504 package versions referenced in your dependency graph. 19,131 (81%) were optimized away.
Dependency resolution completed in 0.041 seconds
Analyzed 4373 packages to resolve dependencies
Analyzed 39521 rules to resolve dependencies
Done, test successful
```

`COMPOSER_POOL_OPTIMIZER=2 composer update -vv`

```
Dependency resolution completed in 0.041 seconds
Analyzed 4373 packages to resolve dependencies
Analyzed 39521 rules to resolve dependencies
Dependency resolution completed in 0.001 seconds
```

This is gonna save a huge amount of time.

PS: I am not rolling back composer now :-P

**Comment by CDRO** [comment]
I've tested this with TYPO3 10 and TYPO3 9 and ran into and exception on finalisation.

I'm aware, that this is probably something that has to be solved on the vendor side rather than composer side, but since the setup for TYPO3 worked in composer 2.1, I'm wondering if snapshot has not introduced an involuntary BC.

The error (with `composer update -vvv`):
```
> pre-autoload-dump: Helhum\Typo3ConsolePlugin\Plugin->listen
Class Cabag\BaloisesessionImport\Command\VipImportCommand located in ./web/typo3conf/ext/baloisesession_import/Classes/Command/OpaccVipImport.php does not comply with psr-4 autoloading standard. Skipping.
Class Cabag\BaloisesessionEvent\Domain\Model\Interview located in ./web/typo3conf/ext/baloisesession_event/Classes/Domain/Model/Inteview.php does not comply with psr-4 autoloading standard. Skipping.
> post-autoload-dump: TYPO3\CMS\Composer\Installer\Plugin->listen
Executing TYPO3 installer scripts
Executing "Helhum\Typo3Console\Composer\InstallerScript\PopulateCommandConfiguration": 
Executing "TYPO3\CMS\Composer\Plugin\Core\InstallerScripts\EntryPoint": 
Executing "TYPO3\CMS\Composer\Plugin\Core\InstallerScripts\EntryPoint": 
Executing "TYPO3\CMS\Core\Composer\CliEntryPoint": 
Executing "TYPO3\CMS\Composer\Plugin\Core\InstallerScripts\EntryPoint": 
> post-autoload-dump: TYPO3\CMS\Composer\Installer\Plugin->listen
> post-autoload-dump: TYPO3\ClassAliasLoader\Plugin->onPostAutoloadDump
> post-autoload-dump: Helhum\Typo3ConsolePlugin\Plugin->listen
> post-autoload-dump: Helhum\Typo3ConsolePlugin\Plugin->listen
> post-autoload-dump: PackageVersions\Installer->dumpVersionsClass
29 packages you are using are looking for funding.
Use the `composer fund` command to find out more!

                                                             
  [RuntimeException]                                         
  Cannot set the alias loader, as it is already registered!  
                                                             

Exception trace:
 () at /Users/st/scripts/projects/project_ci-baloisesession_ch/src/vendor/typo3/class-alias-loader/src/ClassAliasMap.php:72
 TYPO3\ClassAliasLoader\ClassAliasMap::setClassAliasLoader() at /Users/st/scripts/projects/project_ci-baloisesession_ch/src/vendor/typo3/alias-loader-include.php:9
 require() at phar:///usr/local/Cellar/composer/2.1.12/bin/composer/src/Composer/Autoload/AutoloadGenerator.php:1419
 Composer\Autoload\composerRequire() at phar:///usr/local/Cellar/composer/2.1.12/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php:512
 Composer\EventDispatcher\EventDispatcher->getScriptListeners() at phar:///usr/local/Cellar/composer/2.1.12/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php:455
 Composer\EventDispatcher\EventDispatcher->getListeners() at phar:///usr/local/Cellar/composer/2.1.12/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php:172
 Composer\EventDispatcher\EventDispatcher->doDispatch() at phar:///usr/local/Cellar/composer/2.1.12/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php:118
 Composer\EventDispatcher\EventDispatcher->dispatchScript() at phar:///usr/local/Cellar/composer/2.1.12/bin/composer/src/Composer/Installer.php:371
 Composer\Installer->run() at phar:///usr/local/Cellar/composer/2.1.12/bin/composer/src/Composer/Command/UpdateCommand.php:249
 Composer\Command\UpdateCommand->execute() at phar:///usr/local/Cellar/composer/2.1.12/bin/composer/vendor/symfony/console/Command/Command.php:245
 Symfony\Component\Console\Command\Command->run() at phar:///usr/local/Cellar/composer/2.1.12/bin/composer/vendor/symfony/console/Application.php:835
 Symfony\Component\Console\Application->doRunCommand() at phar:///usr/local/Cellar/composer/2.1.12/bin/composer/vendor/symfony/console/Application.php:185
 Symfony\Component\Console\Application->doRun() at phar:///usr/local/Cellar/composer/2.1.12/bin/composer/src/Composer/Console/Application.php:327
 Composer\Console\Application->doRun() at phar:///usr/local/Cellar/composer/2.1.12/bin/composer/vendor/symfony/console/Application.php:117
 Symfony\Component\Console\Application->run() at phar:///usr/local/Cellar/composer/2.1.12/bin/composer/src/Composer/Console/Application.php:128
 Composer\Console\Application->run() at phar:///usr/local/Cellar/composer/2.1.12/bin/composer/bin/composer:73
 require() at /usr/local/Cellar/composer/2.1.12/bin/composer:30

update [--with WITH] [--prefer-source] [--prefer-dist] [--prefer-install PREFER-INSTALL] [--dry-run] [--dev] [--no-dev] [--lock] [--no-install] [--no-autoloader] [--no-scripts] [--no-suggest] [--no-progress] [-w|--with-dependencies] [-W|--with-all-dependencies] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--apcu-autoloader-prefix APCU-AUTOLOADER-PREFIX] [--ignore-platform-req IGNORE-PLATFORM-REQ] [--ignore-platform-reqs] [--prefer-stable] [--prefer-lowest] [-i|--interactive] [--root-reqs] [--] [<packages>]...

```

The same command with composer 2.1.2:
```
> pre-autoload-dump: Helhum\Typo3ConsolePlugin\Plugin->listen
Class Cabag\BaloisesessionImport\Command\VipImportCommand located in ./web/typo3conf/ext/baloisesession_import/Classes/Command/OpaccVipImport.php does not comply with psr-4 autoloading standard. Skipping.
Class Cabag\BaloisesessionEvent\Domain\Model\Interview located in ./web/typo3conf/ext/baloisesession_event/Classes/Domain/Model/Inteview.php does not comply with psr-4 autoloading standard. Skipping.
> post-autoload-dump: TYPO3\CMS\Composer\Installer\Plugin->listen
Executing TYPO3 installer scripts
Executing "Helhum\Typo3Console\Composer\InstallerScript\PopulateCommandConfiguration": 
Executing "TYPO3\CMS\Composer\Plugin\Core\InstallerScripts\EntryPoint": 
Executing "TYPO3\CMS\Composer\Plugin\Core\InstallerScripts\EntryPoint": 
Executing "TYPO3\CMS\Core\Composer\CliEntryPoint": 
Executing "TYPO3\CMS\Composer\Plugin\Core\InstallerScripts\EntryPoint": 
> post-autoload-dump: TYPO3\CMS\Composer\Installer\Plugin->listen
> post-autoload-dump: TYPO3\ClassAliasLoader\Plugin->onPostAutoloadDump
> post-autoload-dump: Helhum\Typo3ConsolePlugin\Plugin->listen
> post-autoload-dump: Helhum\Typo3ConsolePlugin\Plugin->listen
> post-autoload-dump: PackageVersions\Installer->dumpVersionsClass
29 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
> post-update-cmd: Rector\RectorInstaller\Plugin->process
> post-update-cmd: bash scripts/include-fpm.sh
Executing command (CWD): bash scripts/include-fpm.sh
```

I've also reported the issue on the vendors official issue tracking, for good measure:
https://forge.typo3.org/issues/96081

**Reply by Seldaek** [reply]
Please report a new issue for this, let's not turn this discussion thread into a new issue tracker.

**Reply by Seldaek** [reply]
BTW this is most likely due to the change in https://github.com/composer/composer/pull/10065 - doesn't seem related to the poolbuilder. We are now loading `files` autoload, and it appears this package doesn't cope well with this.
